### ISSUE TITLE
OffsetCommitRequests with NULL member_id are generated incorrectly

### ISSUE THREAD

**Issue by justindowning** [oleksiyk/kafka#177]
Occasionally, we see errors like `NoKafkaConnectionError: Kafka server has closed connection`  in the consumer when trying to consume from the broker. 

This appears to be a result of a previous disconnection due to a malformed payload. There is a case where the no-kafka consumer attempts to commit an offset when its member id is NULL. The client protocol incorrectly transforms the NULL value in the OffsetCommitRequest with a length of -1 (signalling a NULL value) and sends the message to the broker. As a result, the broker [throws an error](https://github.com/apache/kafka/blob/bdf4cba047334943aa8357585c4fb379b27e9ffd/clients/src/main/java/org/apache/kafka/common/protocol/types/Type.java#L271-L271) because the OffsetCommitRequest does not allow a NULL value for the consumer member id. 
 
Error on the broker:
```
pri=ERROR t=kafka-network-thread-1-SSL-2 at=Processor Closing socket for x.x.x.x:9096-x.x.x.x:55336 because of error org.apache.kafka.common.errors.InvalidRequestException: Error getting request for apiKey: 8 and apiVersion: 2
Caused by: org.apache.kafka.common.protocol.types.SchemaException: Error reading field 'member_id': String length -1 cannot be negative
```
no-kafka [assumes](https://github.com/oleksiyk/kafka/blob/7a7dfaa204b68a2f406b170865239fed90ce3bab/lib/protocol/common.js#L51-L52) that Kafka's String type and Kafka's NullableString type have the same encoding. The String type in no-kafka incorrectly allows encoding NULL values. As a result, any requests which have a String type (where nulls are not supported) will cause the broker to close the producer or consumer connection. 

The same issue is present with the Bytes type in no-kafka. There is no distinction between Kafka's Bytes type and Kafka's NullableBytes type.

**Comment by oleksiyk** [comment]
I guess you are using Kafka 0.10? In Kafka 0.9 protocol there is no such type as NullableString or NullableBytes

**Comment by oleksiyk** [comment]
Can you point me at protocol specifications for Nullable types? I was relying on a document https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-ProtocolPrimitiveTypes which even in its newer version says the following:

> Variable Length Primitives
bytes, string - These types consist of a signed integer giving a length N followed by N bytes of content. A length of -1 indicates null. string uses an int16 for its size, and bytes uses an int32.

**Comment by justindowning** [comment]
Yes, this is Kafka 0.10. As for the protocol specification, I don't see documention on the wiki you linked to, but I did find the types in the Kafaka Protocol code:
* [NULLABLE_STRING](https://github.com/apache/kafka/blob/bdf4cba047334943aa8357585c4fb379b27e9ffd/clients/src/main/java/org/apache/kafka/common/protocol/types/Type.java#L298-L353)
* [NULLABLE_BYTES](https://github.com/apache/kafka/blob/bdf4cba047334943aa8357585c4fb379b27e9ffd/clients/src/main/java/org/apache/kafka/common/protocol/types/Type.java#L399-L457)

**Comment by oleksiyk** [comment]
I guess the real problem is missing (or null) memberId. Does it happen during a rebalance? 

Changing the protocol in no-kafka is not an option because its a 0.9 client. At least for now.

**Comment by halorgium** [comment]
@oleksiyk in kafka 0.9, there was no support for nullable strings: see https://github.com/apache/kafka/blob/0.9.0.0/clients/src/main/java/org/apache/kafka/common/protocol/types/Type.java#L174-L209

Furthermore, there was no invalid size checks in kafka 0.9: https://issues.apache.org/jira/browse/KAFKA-3426

the wiki page seems to have had the `-1` comment from v2 of the page: https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=30743545
I could believe that this is an unfortunate oversight in documentation. 

no-kafka has always had support for nullable strings which means it generates invalid payloads for both 0.9 and 0.10.

**Comment by oleksiyk** [comment]
If you add some simple logging to the no-kafka protocol (to the 'bytes' type) you'll see how much it does write `-1` as bytes length. And it doesn't fail. You will also see that it **reads** `-1` as well as `0` as its length (particularly the message `key`). I really don't like changing essential parts of the protocol against the protocol specification (wiki) unless I have clear comments from the kafka dev team.

**Comment by oleksiyk** [comment]
I have fixed **reading** null/empty bytes/string/array types in #179 
Previously both `null` (length = -1) and empty (length = 0) values were treated as `null` only.

**Comment by oleksiyk** [comment]
The best solution for original problem (where `memberId` is `null`) is to throw (reject) a call to commit offset if the `memberId` is `null`, e.g. during a rebalance here: https://github.com/oleksiyk/kafka/blob/master/lib/group_consumer.js#L270

**Comment by halorgium** [comment]
@oleksiyk changing the group consumer sounds like a great start. the high level guard would prevent the issue from trickling further downstream.

**Comment by oleksiyk** [comment]
I've published 3.1.8
